packages/sd-jwt/src/action-handler.ts

```diff
@@ -0,0 +1,227 @@
+   * @returns A signed SD-JWT credential.
+   */
+  async createSdJwtVc(args: ICreateSdJwtVcArgs, context: IRequiredContext): Promise<ICreateSdJwtVcResult> {
+    const issuer = args.credentialPayload.iss
+    if (!issuer) {
+      throw new Error('credential.issuer must not be empty')
+    }
+
+    const { alg, key } = await this.getSignKey(issuer, context)
+
+    //TODO: let the user also insert a method to sign the data
+    const signer: Signer = async (data: string) => context.agent.keyManagerSign({ keyRef: key.kid, data })
+
+    const sdjwt = new SDJwtVcInstance({
+      signer,
+      hasher: this.algorithms.hasher,
+      saltGenerator: this.algorithms.saltGenerator,
+      signAlg: alg,
+      hashAlg: 'SHA-256',
+    })
+
+    const credential = await sdjwt.issue(
+      args.credentialPayload,
+      args.disclosureFrame as DisclosureFrame<typeof args.credentialPayload>,
+    )
+    return { credential }
+  }
+
+  /**
+   * Get the key to sign the SD-JWT
+   * @param issuer - did url like did:exmaple.com#key-1
+   * @param context - agent instance
+   * @returns the key to sign the SD-JWT
+   */
+  private async getSignKey(issuer: string, context: IRequiredContext) {
+    const identifier = await context.agent.didManagerGet({
+      did: issuer.split('#')[0],
+    })
+    const doc = await mapIdentifierKeysToDocWithJwkSupport(identifier, 'assertionMethod', context)
+    if (!doc || doc.length === 0) throw new Error('No key found for signing')
+    const key = doc.find((key) => key.meta.verificationMethod.id === issuer)
+    if (!key) throw new Error('No key found with the given id')
+    let alg: string
+    //transform the key type to the alg
+    switch (key.type) {
+      case 'Ed25519':
+        alg = 'EdDSA'
+        break
+      case 'Secp256k1':
+        alg = 'ES256K'
+        break
+      case 'Secp256r1':
+        alg = 'ES256'
+        break
+      default:
+        throw new Error(`unsupported key type ${key.type}`)
+    }
+    return { alg, key }
+  }
+
+  /**
+   * Create a signed SD-JWT presentation.
+   * @param args - Arguments necessary for the creation of a SD-JWT presentation.
+   * @param context - This reserved param is automatically added and handled by the framework, *do not override*
+   * @returns A signed SD-JWT presentation.
+   */
+  async createSdJwtVcPresentation(args: ICreateSdJwtVcPresentationArgs, context: IRequiredContext): Promise<ICreateSdJwtVcPresentationResult> {
+    const cred = await SDJwt.fromEncode(args.presentation, this.algorithms.hasher)
+    const claims = await cred.getClaims<Claims>(this.algorithms.hasher)
+    let holderDID: string
+    // we primarly look for a cnf field, if it's not there we look for a sub field. If this is also not given, we throw an error since we can not sign it.
+    if (claims.cnf?.jwk) {
+      const key = claims.cnf.jwk
+      holderDID = `did:jwk:${encodeJoseBlob(key)}#0`
```

Comment:
Hmzz, this seems a bit naive. We probably should place a TODO saying that we should convert the JWK to hex and search for the appropriate key and associated DID. Certainly doesn't apply to did:jwk only, as you can represent any DID key as a JWK. So whenever you encounter a JWK it doesn't mean it had to come from a did:jwk in the system. It just can always be represented as a did:jwk

Reply:
added a TODO